Reduce file system calls during test setup.

dragomirtitian opened the pull request:
Improvements to reduce fs calls during test discovery:

1. Use `opendirSync` to read the directory content. This reduces the need to call `fs.existsSync` and `fs.statSync`
2. Cache `enumerateTestFiles` result to avoid having to enumerating the files again
3. Delay creation of vfs for project tests (as creating the vfs also mounts real folders which also touch the file system)

Improvements in test discovery time (anecdotal):

1. Linux node 18 debug - ~1.4 seconds -> ~800ms
2. Windows 10 node 20 debug - 8s -> 2s

Performance of running all the tests does not appear to be impacted as the cost of test discovery is not a big percentage of running all tests and the work that si delayed will have to eventually be executed anyway.

Fixes #56790

jakebailey commented on src/harness/harnessIO.ts:
```diff
@@ -87,19 +87,20 @@ function createNodeIO(): IO {
     function listFiles(path: string, spec: RegExp, options: { recursive?: boolean; } = {}) {
         function filesInFolder(folder: string): string[] {
             let paths: string[] = [];
+            const dir = fs.opendirSync(folder);
```
Could we just use `readdirSync` with `withFileTypes: true` instead? With recursion, we'd be holding directories open the entire way down, but maybe we never go deep enough.

(I'd also say that readdir has `recursive` now, but it's only available in Node 18+)

dragomirtitian replied:
That is a fair point. I don't think the we have a very deep directory structure in tests so it's probably fine.  I will read the whole current directory first close it and then continue reading sub directories just to be safe.

jakebailey replied:
Is withFileTypes enough to get all of the needed info? That's effectively what loading it all at once does.

I'm not against holding them open really, the code is plenty clear and I'm sure fine as is; ignore my hold open comment (speaking aloud too much)

dragomirtitian replied:
I changed it over to use withFileTypes

robpalme commented on src/harness/harnessIO.ts:
```diff
@@ -87,19 +87,20 @@ function createNodeIO(): IO {
     function listFiles(path: string, spec: RegExp, options: { recursive?: boolean; } = {}) {
         function filesInFolder(folder: string): string[] {
             let paths: string[] = [];
+            const dir = fs.opendirSync(folder);
```
Given we're optimizing for performance, does increasing the default buffer size from 32 help? That option has been available since Node 12.

dragomirtitian replied:
Switched over to withFileTypes

jakebailey replied:
Thanks; does the perf improvement change at all?

dmichon-msft commented on src/harness/harnessIO.ts:
```diff
@@ -87,19 +87,18 @@ function createNodeIO(): IO {
     function listFiles(path: string, spec: RegExp, options: { recursive?: boolean; } = {}) {
         function filesInFolder(folder: string): string[] {
             let paths: string[] = [];
+            for (const entry of fs.readdirSync(folder, { withFileTypes: true })) {
```
Apply the same optimization to `getAccessibleFileSystemEntries`?

jakebailey replied:
Yeah, this is a good idea; the implementation we have in `sys.ts` already uses `withFileTypes`.

(That being said, this whole block just duplicates functionality in `sys.ts` so I do wonder how much copy pasting we could avoid.)

sheetalkamat commented on src/testRunner/compilerRunner.ts:
```diff
@@ -51,9 +51,10 @@ export class CompilerBaselineRunner extends RunnerBase {
     public enumerateTestFiles() {
         // see also: `enumerateTestFiles` in tests/webTestServer.ts
-        return this.enumerateFiles(this.basePath, /\.tsx?$/, { recursive: true }).map(CompilerTest.getConfigurations);
+        return this.loadedTestFiles ??= this.enumerateFiles(this.basePath, /\.tsx?$/, { recursive: true }).map(CompilerTest.getConfigurations);
```
I think these need to be released at some point as otherwise we are holding onto this data for too long (till all tests are running?)

jakebailey replied:
It would be until the process ends I believe, but I don't think that this actually ends up holding on to too much; it's just a list of filenames, right?

sheetalkamat replied:
Its typed :  `CompilerFileBasedTest` so not sure if its just file names or more ?

jakebailey replied:
So, it'll include filenames, but also file contents and configurations. Honestly, I'm shocked that we do that stuff up front, but I suppose you need the file contents to know the config to then know what tests to generate...

dragomirtitian replied:
@sheetalkamat @jakebailey The more aggressive approach would be to return just the `string[]` from `enumerateTestFiles`. I do not see an instance where the extra information returned is used. Maybe it was in the past and it is no longer needed? https://github.com/microsoft/TypeScript/commit/fd8b18244bffb349cd1bf54e31e597041422d24e

This would remove the double reading and parsing of the test files and there would not be a need to cache the result.

jakebailey replied:
That actually seems great.

dragomirtitian replied:
Pushed the new version.

jakebailey commented:
Looks good, though there's still an open thread about another function that could use `withFileTypes`.

That being said, I also don't see as great of an improvement as your benchmark; before:

```
Benchmark 1: hereby runtests -t "notarealtest" --no-typecheck --no-lint
  Time (mean ± σ):      2.265 s ±  0.012 s    [User: 2.540 s, System: 0.652 s]
  Range (min … max):    2.250 s …  2.286 s    10 runs
```

After:

```
Benchmark 1: hereby runtests -t "notarealtest" --no-typecheck --no-lint
  Time (mean ± σ):      2.093 s ±  0.039 s    [User: 2.351 s, System: 0.559 s]
  Range (min … max):    2.035 s …  2.155 s    10 runs
```

8% faster is still good, but you had seen 75% faster...

sheetalkamat reviewed (commented):
This looks ok to me except @jakebailey s suggestion on just replacing `listFiles` and `getAccessibleFileSystemEntries` to use `ts.sys.getAccessibleFileSystemEntries`

dragomirtitian commented:
> This looks ok to me except @jakebailey s suggestion on just replacing `listFiles` and `getAccessibleFileSystemEntries` to use `ts.sys.getAccessibleFileSystemEntries`

@sheetalkamat One small problem is that `getAccessibleFileSystemEntries` is not exposed in `sys`. I exposed it as internal. hope that is ok. I rewrote `listFiles` to use the sys version and also removed the local implementation of `getAccessibleFileSystemEntries`

jakebailey commented on src/harness/harnessIO.ts:
```diff
@@ -170,7 +143,7 @@ function createNodeIO(): IO {
         exit: exitCode => ts.sys.exit(exitCode),
         readDirectory: (path, extension, exclude, include, depth) => ts.sys.readDirectory(path, extension, exclude, include, depth),
-        getAccessibleFileSystemEntries,
+        getAccessibleFileSystemEntries: ts.sys.getAccessibleFileSystemEntries!,
```
I don't think that it matters, but every other function in this list is an arrow function into `sys`, not the actual value.
